// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package privacy_sandbox.server_common;

message TelemetryConfig {
  enum TelemetryMode {
    UNKNOWN_TYPE = 0;

    // No telemetry (metrics/traces).
    OFF = 1;

    // Production mode.
    // `Privacy::kImpacting` metrics must be noised.
    PROD = 2;

    // Same as PROD in prod build;
    // In experiment build, enables metrics without noise; and tracing is also
    // enabled.
    EXPERIMENT = 3;

    // Same as PROD in prod build;
    // In experiment build, exports both noised and raw data for
    // `Privacy::kImpacting` metrics.
    COMPARE = 4;
  }
  TelemetryMode mode = 1;

  // If non-empty, only output these metrics in both here and default list.
  // If empty, the default list of metrics is used.
  repeated MetricConfig metric = 2;

  // The time interval between two consecutive metric exports.
  int32 metric_export_interval_ms = 3;

  // The time interval between two consecutive differential privacy aggregated.
  // metric exports, must be at least metric_export_interval_ms
  int32 dp_export_interval_ms = 4;
}

message MetricConfig {
  // name should match `metric::Definition` in the list of server's metric
  // provided when creating `ContextMap`
  string name = 1;
}
